engine/lite/components/request-response-tests/src/test/scala/pl/touk/nussknacker/engine/lite/components/requestresponse/CollectTransformerTest.scala

```diff
@@ -0,0 +1,137 @@
+package pl.touk.nussknacker.engine.lite.components.requestresponse
+
+import akka.http.scaladsl.model.{HttpMethods, HttpRequest}
+import cats.data.{NonEmptyList, ValidatedNel}
+import io.circe.Json
+import org.scalatest.Inside.inside
+import org.scalatest.LoneElement._
+import org.scalatest.OptionValues
+import org.scalatest.funsuite.AnyFunSuite
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.prop.TableDrivenPropertyChecks
+import pl.touk.nussknacker.engine.api.context.ProcessCompilationError
+import pl.touk.nussknacker.engine.api.context.ProcessCompilationError.ExpressionParserCompilationError
+import pl.touk.nussknacker.engine.build.{ProcessGraphBuilder, ScenarioBuilder}
+import pl.touk.nussknacker.engine.canonicalgraph.CanonicalProcess
+import pl.touk.nussknacker.engine.lite.api.commonTypes.ErrorType
+import pl.touk.nussknacker.engine.lite.components.requestresponse.jsonschema.sinks.JsonRequestResponseSink.SinkRawEditorParamName
+import pl.touk.nussknacker.engine.lite.util.test.RequestResponseTestScenarioRunner._
+import pl.touk.nussknacker.engine.requestresponse.api.openapi.RequestResponseOpenApiSettings.{
+  InputSchemaProperty,
+  OutputSchemaProperty
+}
+import pl.touk.nussknacker.engine.spel.Implicits._
+import pl.touk.nussknacker.engine.util.test.TestScenarioRunner
+import pl.touk.nussknacker.test.{EitherValuesDetailedMessage, ValidatedValuesDetailedMessage}
+
+class CollectTransformerTest
+    extends AnyFunSuite
+    with Matchers
+    with EitherValuesDetailedMessage
+    with OptionValues
+    with TableDrivenPropertyChecks
+    with ValidatedValuesDetailedMessage {
+
+  private val inputSchema =
+    """
+      |{
+      |  "type": "array",
+      |  "items": {
+      |    "type": "number"
+      |  }
+      |}
+      |""".stripMargin
+
+  private val outputSchema =
+    """
+      |{
+      |  "type": "array",
+      |  "items": {
+      |    "type": "string"
+      |  }
+      |}
+      |""".stripMargin
+
+  test("should collect elements after for-each") {
+    val scenario = scenarioBuilderWithSchemas
+      .customNode("for-each", "outForEach", "for-each", "Elements" -> "#input")
+      .buildSimpleVariable("someVar", "ourVar", "'x = ' + (#outForEach * 2)")
+      .customNode("collect", "outCollector", "collect", "Input expression" -> "#ourVar")
+      .emptySink("response", "response", SinkRawEditorParamName -> "true", "Value" -> "#outCollector")
```

Comment:
How user can access e.g. #input.someOtherField in the proposed approach? As I understand, it worked before your change?

Reply:
He can't. All variables need to be put in collect node so they can be used later. https://staging.nussknacker.io/visualization/ljd-collect-context-bug shows a scenario that tries to access a variable defined before collect node. Currently NPE is thrown but the variable should not be available for user.

But, as you said, we should rethink how 'for-each' and 'collect' are supposed to work in request-response mode.